### PR TITLE
Fixing the state aggregation system to properly handle requisities

### LINKED ISSUE #47628: State Aggregate Breaks Requisites
Setting ```aggregate: True``` (https://docs.saltstack.com/en/latest/ref/states/aggregate.html) for ```pkg``` can significantly reduce run time.  However, any requisites attached to pkg states seem to be lost.

**Package Repositories:** If, for example, a package requires a repository to be installed first, the package install can fail because it can get moved ahead of the repository install.  Neither setting the package install to ```require``` the repository install, nor setting the repository install to ```require_in``` the package install seems to help.  A workaround that does appear to succeed is setting ```order: 1``` on the repository install.

**Services:** Another example of an issue is when a package install requires a service restart.  The service won't get restarted because any requisite ties between the package install and the service are lost when ```aggregate``` is used.

This is related to #21312, which has been marked as stale, but the issue appears to persist.

### PR DESCRIPTION
**Pull request by garethgreenaway**
### What does this PR do?
Fixing the state aggregation system to properly handle requisities.
Fixing pkg state to exclude packages from aggregation if the hold attribute is in the state. 
Moving tests/unit/test_state.py to pytest.

### What issues does this PR fix or reference?
Fixes: #47628 

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [ ] Docs
- [ ] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [ ] Tests written/updated

### Commits signed with GPG?
Yes/No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

**Comment by garethgreenaway** [comment]
I did notice one change to make with the move to pytest, need to remove one instance of RUNTIME_VARS.

File: salt/state.py [hunk 1]
```diff
@@ -841,6 +841,35 @@ def _mod_init(self, low):
...
                 self.mod_init.add(low["state"])
 
+    def _aggregate_requisites(self, low, chunks):
+        """
+        Aggregate the requisites
+        """
+        requisites = {}
+        low_state = low["state"]
+        for chunk in chunks:
+            log.debug("==== _aggregate_requisites - chunk %s ===", chunk)
```
Comment (Ch3LL) [inline salt/state.py:851]: Does this need to be removed?

File: salt/state.py [hunk 2]
```diff
@@ -841,6 +841,35 @@ def _mod_init(self, low):
...
+                            requisites[req].extend(chunk[req])
+                        else:
+                            requisites[req] = chunk[req]
+                for req in STATE_REQUISITE_IN_KEYWORDS:
+                    if req in chunk:
+                        if req in requisites:
+                            requisites[req].extend(chunk[req])
+                        else:
+                            requisites[req] = chunk[req]
+        log.debug("==== _aggregate_requisites - requisites %s ===", requisites)
```
Comment (Ch3LL) [inline salt/state.py:869]: here too

File: salt/state.py [hunk 3]
```diff
@@ -857,6 +886,8 @@ def _mod_aggregate(self, low, running, chunks):
             if agg_fun in self.states:
                 try:
                     low = self.states[agg_fun](low, chunks, running)
+                    log.debug("==== after mod_aggregate - low %s ===", low)
```
Comment (Ch3LL) [inline salt/state.py:889]: here as well

File: salt/state.py [hunk 4]
```diff
@@ -891,6 +922,7 @@ def _run_check(self, low_data):
         if "onlyif" in low_data:
             _ret = self._run_check_onlyif(low_data, cmd_opts)
             ret["result"] = _ret["result"]
+
```
Comment (Ch3LL) [inline salt/state.py:925]: remove new line

File: salt/states/pkg.py [hunk 5]
```diff
@@ -3486,6 +3486,7 @@ def mod_aggregate(low, chunks, running):
         "removed",
         "purged",
     ]
+    log.debug("=== low %s chunks %s running %s ===", low, chunks, running)
```
Comment (Ch3LL) [inline salt/states/pkg.py:3489]: remove here too

File: salt/states/pkg.py [hunk 6]
```diff
@@ -3511,25 +3512,28 @@ def mod_aggregate(low, chunks, running):
...
+                                pkgs.append({chunk["name"]: version})
+                            else:
+                                pkgs.append(chunk["name"])
+                            chunk["__agg__"] = True
     if pkg_type is not None and pkgs:
         if pkg_type in low:
             low[pkg_type].extend(pkgs)
         else:
             low[pkg_type] = pkgs
+    log.debug("=== in mod_aggregate - low %s ===", low)
```
Comment (Ch3LL) [inline salt/states/pkg.py:3536]: here too

**Review by Ch3LL — CHANGES_REQUESTED** [review]

**Review by Ch3LL — DISMISSED** [review]

File: salt/state.py [hunk 7]
```diff
@@ -838,6 +838,33 @@ def _mod_init(self, low):
...
+            # and __agg__ is True, add the requisites from the
+            # chunk to those in the low.
+            if chunk["state"] == low["state"] and chunk.get("__agg__"):
+                for req in STATE_REQUISITE_KEYWORDS:
+                    if req in chunk:
+                        if req in requisites:
+                            requisites[req].extend(chunk[req])
+                        else:
+                            requisites[req] = chunk[req]
+                for req in STATE_REQUISITE_IN_KEYWORDS:
```
Comment (waynew) [inline salt/state.py:859]: **question**: is it requisite (har har) that we loop over these independently, or would `for req in STATE_REQUISITE_KEYWORDS + STATE_REQUISITE_IN_KEYWORDS` (or `itertools.chain(...)`) be clearer here?

File: salt/states/pkg.py [hunk 8]
```diff
@@ -3528,20 +3528,22 @@ def mod_aggregate(low, chunks, running):
...
-                        pkgs.extend(chunk["pkgs"])
-                        chunk["__agg__"] = True
-                    elif "name" in chunk:
-                        version = chunk.pop("version", None)
-                        if version is not None:
-                            pkgs.append({chunk["name"]: version})
-                        else:
-                            pkgs.append(chunk["name"])
-                        chunk["__agg__"] = True
+                # If hold exists in the chunk, do not add to aggregation
```
Comment (waynew) [inline salt/states/pkg.py:3531]: **question** this describes the what - can we describe the why?

**Review by waynew — CHANGES_REQUESTED** [review]
I'd like to also check the test code (I'm assuming that most of this was the UnitTest->pytest conversion, so that stuff is probably fairly solid...) but I figured I could get these questions in now.

**Review by waynew — APPROVED** [review]
:+1:

**Pull request merged**